src/derive_utils.rs

```diff
@@ -34,77 +34,74 @@ pub struct ParamDescription {
-                        "Argument given by name ('{}') and position ({})",
-                        p.name,
-                        i + 1
-                    )));
+                    raise_error!("got multiple values for argument: {}", p.name)
                 }
+                kwargs.as_ref().unwrap().del_item(p.name).unwrap();
+                Some(kwarg)
             }
             None => {
                 if p.kw_only {
                     if !p.is_optional {
-                        return Err(TypeError::py_err(format!(
-                            "Required argument ('{}') is keyword only argument",
-                            p.name
-                        )));
+                        raise_error!("missing required keyword-only argument: {}", p.name)
                     }
-                    *out = None;
+                    None
                 } else if i < nargs {
-                    *out = Some(args.get_item(i));
+                    used_args += 1;
+                    Some(args.get_item(i))
                 } else {
-                    *out = None;
                     if !p.is_optional {
-                        return Err(TypeError::py_err(format!(
-                            "Required argument ('{}') (pos {}) not found",
-                            p.name,
-                            i + 1
-                        )));
+                        raise_error!("missing required positional argument: {}", p.name)
                     }
+                    None
                 }
             }
         }
     }
-    if !accept_kwargs && used_keywords != nkeywords {
-        // check for extraneous keyword arguments
-        for item in kwargs.unwrap().items().iter() {
-            let item = <PyTuple as PyTryFrom>::try_from(item)?;
-            let key = <PyString as PyTryFrom>::try_from(item.get_item(0))?.to_string()?;
-            if !params.iter().any(|p| p.name == key) {
-                return Err(TypeError::py_err(format!(
-                    "'{}' is an invalid keyword argument for this function",
-                    key
-                )));
-            }
-        }
+    let is_kwargs_empty = kwargs.as_ref().map_or(true, |dict| dict.is_empty());
+    // Raise an error when we get an unknown key
+    if !accept_kwargs && !is_kwargs_empty {
+        let (key, _) = kwargs.unwrap().iter().next().unwrap();
+        raise_error!("got an unexpected keyword argument: {}", key)
+    }
+    // Raise an error when we get too many positional args
+    if !accept_args && used_args < nargs {
+        raise_error!(
+            "takes at most {} positional argument{} ({} given)",
+            used_args,
+            if used_args == 1 { "" } else { "s" },
+            nargs
+        )
+    }
+    // Adjust the remaining args
+    if accept_args {
+        let slice = args
+            .slice(used_args as isize, nargs as isize)
+            .into_object(py);
+        *args = py.checked_cast_as(slice).unwrap();
```

Comment:
Why do we have to truncate the args?

Reply:
Well, this PR assumes that the semantics on the Rust side is supposed to mimic the semantics on the Python side.

I.e. when you declare a parameter as `*args`, it only gets those positional arguments that are not matched by other parameters.

Same for keywords; the dictionary Python gives you for `**kwds` parameters does not contain keywords that are already matched elsewhere.  Python obviously creates fresh dictionaries for this purpose, e.g.

```
>>> def f(**kwds):
...  kwds.clear()
... 
>>> a = {'a': 1}
>>> f(**a)
>>> a
{'a': 1}
```